### PR TITLE
Dash WO teacher dashboard prototype

### PR DESCRIPTION
**Pull request by bradley-erickson**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -22,4 +22,6 @@ learning_observer/learning_observer/static_data/repos/
 learning_observer/learning_observer/static_data/courses.json
 learning_observer/learning_observer/static_data/students.json
 learning_observer/passwd.lo
---*
\ No newline at end of file
+--*
+.venv/
+learning_observer/prototypes/dash_wo_teacher_dashboard/data/
```
Comment (pmitros) [inline .gitignore:27]: I'd consider having a generic line / opt-out for all data files we don't want to commit.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/aiohttp_server.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from aiohttp import web
+from aiohttp_wsgi import WSGIHandler
+from app import app
+
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/aiohttp_server.py:4]: This PR should be run through `pycodestyle` and `pylint`. We should follow the former. The latter makes helpful suggestions, but doesn't need to be followed 100% of the time

File: learning_observer/prototypes/dash_wo_teacher_dashboard/app.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+# local imports
+from components import navbar
+
+app = dash.Dash(
+    __name__,
+    use_pages=True,
+    external_stylesheets=[
+        dbc.themes.MINTY, # bootstrap styling
+        dbc.icons.FONT_AWESOME, # icons
+        'https://cdn.jsdelivr.net/gh/AnnMarieW/dash-bootstrap-templates@V1.0.6/dbc.min.css', # dcc bootstrap styling
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/app.py:15]: FYI: We should grab these locally when running in LO. CDNs can introduce security issues:

1. File changes to be hostile code on the CDN
2. CDN spies on users

We have code for grabbing and authenticating in the LO repo. This is fine for development (as here; we're not going to call `app.py` from LO).

File: learning_observer/prototypes/dash_wo_teacher_dashboard/app.py [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+)
+
+# create layout
+def serve_layout():
+    return html.Div(
+        [
+            navbar,
+            dbc.Container(
+                dash.page_container,
+                class_name='my-2',
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/app.py:28]: Meaningful class name?

I'd also consider moving the layouts out of `app.py`, so we can import them without making an app.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/assets/highlight.css [hunk 5]
```diff
@@ -0,0 +1,56 @@
+/* main highlights */
+.main {
+    --bs-bg-opacity: 0.5;
+    background-color: rgba(var(--bs-success-rgb), var(--bs-bg-opacity)) !important;
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/assets/highlight.css:4]: Do we need all the `!important`? 

re: Phone conversation. If this is only used for examples, it should at least be commented that it's not used in the main code.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/assets/scripts.js [hunk 6]
```diff
@@ -0,0 +1,115 @@
+if (!window.dash_clientside) {
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/assets/scripts.js:1]: Comment what this is.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/app.py (hunk 4 above)
Comment (pmitros) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/app.py:28]: re: Phone conversation. Add comment:

     # Bootstrap margin

File: learning_observer/prototypes/dash_wo_teacher_dashboard/assets/styles.css [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+  }
+
+
+a.disabled {
+    color: var(--bs-gray);
+    text-decoration: none;
+    pointer-events: none;
+}
+
+.shadow-card:hover {
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/assets/styles.css:33]: I'd love to have comments indicating what UX element these correspond to, or what their purpose is.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py [hunk 8]
```diff
@@ -0,0 +1,19 @@
+# package imports
+from dash import html
+import dash_bootstrap_components as dbc
+import datetime
+import platform
+import random
+
+date_format = '%B %#d, %G' if platform.system() == 'Windows' else '%B %-d, %G'
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py:8]: This is defined and never used. What is this?

File: learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py [hunk 9]
```diff
@@ -0,0 +1,19 @@
...
+# as long as we don't have callbacks involve the
+# functions could easily be turned into class methods
+class Assignment:
+    def __init__(self, id, name, description):
+        self.id = id
+        self.name = name
+        self.description = description
+        self.essay_type = random.choice(['Narrative', 'Argumentative', 'Other'])
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py:16]: Not for this PR, but it'd be helpful to think through this.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/components/course.py [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+    def __init__(self, id, role):
+        self.id = id
+        self.name = f'Course {id}'
+        self.role = role
+
+        if role == 'teacher':
+            self.students = self.fetch_students()
+            self.assignments = [Assignment(0, assignment['name'], assignment['description'])]
+        elif role == 'student':
+            self.dashboard = 'BRUH'
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/components/course.py:26]: What is BRUH?

Comment (pmitros) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/components/course.py:26]: We don't have (or want) roles at the stage, and that's not how you handle them. For now, everyone's a teacher.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/components/navbar.py [hunk 11]
```diff
@@ -0,0 +1,22 @@
+# package imports
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/components/navbar.py:1]: File-level comments would be helpful (everywhere) defining what files are.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/data_preprocessing.py [hunk 12]
```diff
@@ -0,0 +1,188 @@
+# package imports
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/data_preprocessing.py:1]: Do you want this file committed in the PR? This seems like a one-off script. Or if so, do you want it in the main directory? It also seems to cut-and-paste old AWE Workbench code. Shouldn't this be imported?

File: learning_observer/prototypes/dash_wo_teacher_dashboard/examples/drag.py [hunk 13]
```diff
@@ -0,0 +1,79 @@
+'''
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/examples/drag.py:1]: We spoke by phone, and you asked what to do with these. As I'm getting here, I'm okay either way, but I think I'd consider having these, one-off scripts, etc. together, with a README file explaining what they are.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/layout.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+from .teacher_dashboard import create_teacher_dashboard
+
+dash.register_page(
+    __name__,
+    path_template='/dashboard',
+    title='Dashboard'
+)
+
+def layout():
+    role = 'teacher'
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/layout.py:16]: Please kill roles. At some point, we can discuss good v. bad placeholder code.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py [hunk 15]
```diff
@@ -0,0 +1,420 @@
+# package imports
+from dash import html, dcc, clientside_callback, ClientsideFunction, Output, Input, State, ALL
+import dash_bootstrap_components as dbc
+
+prefix = 'teacher-dashboard-settings'
+# settings ids
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py:6]: A few more words would be appreciated....

File: learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py [hunk 16]
```diff
@@ -0,0 +1,420 @@
...
+show_hide_settings_text_collapse = f'{prefix}-show-hide-text-collapse'
+show_hide_settings_text_radioitems = f'{prefix}-show-hide-text-radioitems'
+
+show_hide_settings_highlight_collapse = f'{prefix}-show-hide-highlight-collapse'
+show_hide_settings_highlight_checklist = f'{prefix}-show-hide-highlight-radioitems'
+
+show_hide_settings_indicator_collapse = f'{prefix}-show-hide-indicator-collapse'
+show_hide_settings_indicator_checklist = f'{prefix}-show-hide-indicator-checklist'
+
+open_btn = dbc.Button(
```
Comment (pmitros) [inline learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py:30]: As a style issue, I might consider `([` without the newline where possible. You'll have much less indent that way.

**Review by pmitros — COMMENTED** [review]
Overall, I'd consider commenting on much more of the HTML / CSS, and assuming people don't know dash, dbc, class names, etc.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py (hunk 16 above)
Comment (bradley-erickson) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py:30]: I disagree with this. If the child list is the attribute, then the `([` looks nice; however, if we add additional attributes to the Element, then we can't nicely close the `])`.

I try to always follow the last piece of code shown below.
```python
# only children
html.Div([
    'Text of child 1',
    'Text of child 2'
])
# additional attribute options
html.Div([
    'Text of child 1',
    'Text of child 2'
    ],
    id='example'
)
html.Div([
        'Text of child 1',
        'Text of child 2'
    ],
    id='example'
)
# how it currently is,
html.Div(
    [
        'Text of child 1',
        'Text of child 2'
    ],
    id='example'
)
```

File: learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py (hunk 8 above)
Comment (bradley-erickson) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/components/assignment.py:8]: At one point the Assignment included the date. This was used to represent the date in Month Day, Year text format. Windows uses slightly different cases.

File: learning_observer/prototypes/dash_wo_teacher_dashboard/data_preprocessing.py (hunk 12 above)
Comment (bradley-erickson) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/data_preprocessing.py:1]: It's a single data-reconstruction method that's copy'd from it. I just needed it for processing the datasets (the ones you shared). This creates the dataset that is read into the current UI

File: learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py (hunk 16 above)
Comment (pmitros) [reply learning_observer/prototypes/dash_wo_teacher_dashboard/pages/assignment_page/settings.py:31]: Fair enough. We can have stylistic differences. I'll mark as resolved.

**Comment by pmitros** [comment]
@bradley-erickson I'd be inclined to merge and do further clean-ups later. Sound good?

I'm displaying `dash` dashboards on my end, and it'd be nice to get the two working together ASAP. A lot more clean-up can also happen during integration.

**Pull request merged**